From 8ff7380c385b163cc8183b988fdb2dcb8196006f Mon Sep 17 00:00:00 2001
From: Jenny TC <jenny.tc@intel.com>
Date: Fri, 9 Jan 2015 14:16:08 +0530
Subject: [PATCH 1/3] wcove_pmic: Read usb compliance bit

Read USB compliance bit from OEM1 table and set pmic_ccsm
pdata to indicate the same

Change-Id: I550af3b02642b74d0f51ffda818a9d24a0623528
Signed-off-by: Jenny TC <jenny.tc@intel.com>
---
 drivers/mfd/intel_soc_pmic_wc.c       | 20 ++++++++++++++++++++
 include/linux/power/intel_pmic_ccsm.h |  1 +
 2 files changed, 21 insertions(+)

diff --git a/drivers/mfd/intel_soc_pmic_wc.c b/drivers/mfd/intel_soc_pmic_wc.c
index fba04e2..ab4c43f 100644
--- a/drivers/mfd/intel_soc_pmic_wc.c
+++ b/drivers/mfd/intel_soc_pmic_wc.c
@@ -34,6 +34,7 @@
 #include <linux/iio/types.h>
 #include <linux/power/intel_pmic_ccsm.h>
 #include <linux/mfd/intel_soc_pmic_wcove.h>
+#include <asm/intel_em_config.h>
 #include "./intel_soc_pmic_core.h"
 
 #define WHISKEY_COVE_IRQ_NUM	17
@@ -60,6 +61,8 @@
 #define MTHRMIRQ3	0xDA
 #define MCHGRIRQ	0x17
 
+#define FPO0_USB_COMP_OFFSET 0x01
+
 static bool wcove_init_done;
 
 static struct gpadc_regmap_t whiskeycove_gpadc_regmaps[GPADC_NUM_CHANNELS] = {
@@ -581,6 +584,21 @@ err:
 	return ret;
 }
 
+static bool is_usb_compliant_charging(void)
+{
+	struct em_config_oem1_data em_config;
+
+
+	/* On error return true to keep usb compliance and charge
+	*  with 100mA until enumerated.
+	*/
+	if (em_config_get_oem1_data(&em_config) < 0)
+		return true;
+
+	/* 0 - usb compliance, 1 - no usb compliance */
+	return !(em_config.fpo_0 & FPO0_USB_COMP_OFFSET);
+}
+
 static void wcove_set_ccsm_config(void)
 {
 	static struct intel_pmic_ccsm_platform_data pdata;
@@ -599,6 +617,8 @@ static void wcove_set_ccsm_config(void)
 
 	intel_soc_pmic_set_pdata("wcove_ccsm", &pdata,
 		sizeof(pdata), 0);
+
+	pdata.usb_compliance = is_usb_compliant_charging();
 }
 
 static void wcove_set_bcu_pdata(void)
diff --git a/include/linux/power/intel_pmic_ccsm.h b/include/linux/power/intel_pmic_ccsm.h
index 9240706..edb52cc 100644
--- a/include/linux/power/intel_pmic_ccsm.h
+++ b/include/linux/power/intel_pmic_ccsm.h
@@ -99,6 +99,7 @@ struct intel_pmic_ccsm_platform_data {
 	struct pmic_regs *reg_map;
 	struct pmic_ccsm_int_cfg *intmap;
 	int intmap_size;
+	bool usb_compliance;
 };
 
 extern int intel_pmic_get_status(void);
-- 
1.9.1

